// @sanity/sdk-react/src/_exports/sdk-react.ts
/**
 * @module exports
 */
export {AuthBoundary, type AuthBoundaryProps} from '../components/auth/AuthBoundary'
export {SanityApp, type SanityAppProps} from '../components/SanityApp'
export {SDKProvider, type SDKProviderProps} from '../components/SDKProvider'
export {ComlinkTokenRefreshProvider} from '../context/ComlinkTokenRefresh'
export {renderSanityApp} from '../context/renderSanityApp'
export {ResourceProvider, type ResourceProviderProps} from '../context/ResourceProvider'
export {
  SanityInstanceProvider,
  type SanityInstanceProviderProps,
} from '../context/SanityInstanceProvider'
export {SDKStudioContext, type StudioWorkspaceHandle} from '../context/SDKStudioContext'
export {
  useAgentGenerate,
  useAgentPatch,
  useAgentPrompt,
  useAgentTransform,
  useAgentTranslate,
} from '../hooks/agent/agentActions'
export {
  type AgentResourceContextOptions,
  useAgentResourceContext,
} from '../hooks/agent/useAgentResourceContext'
export {useAuthState} from '../hooks/auth/useAuthState'
export {useAuthToken} from '../hooks/auth/useAuthToken'
export {useCurrentUser} from '../hooks/auth/useCurrentUser'
export {useDashboardOrganizationId} from '../hooks/auth/useDashboardOrganizationId'
export {useHandleAuthCallback} from '../hooks/auth/useHandleAuthCallback'
export {useLoginUrl} from '../hooks/auth/useLoginUrl'
export {useLogOut} from '../hooks/auth/useLogOut'
export {useVerifyOrgProjects} from '../hooks/auth/useVerifyOrgProjects'
export {useClient} from '../hooks/client/useClient'
export {
  type FrameConnection,
  type FrameMessageHandler as MessageHandler,
  useFrameConnection,
  type UseFrameConnectionOptions,
} from '../hooks/comlink/useFrameConnection'
export {
  useWindowConnection,
  type UseWindowConnectionOptions,
  type WindowConnection,
  type WindowMessageHandler,
} from '../hooks/comlink/useWindowConnection'
export {useSanityInstance} from '../hooks/context/useSanityInstance'
export {useDashboardNavigate} from '../hooks/dashboard/useDashboardNavigate'
export {useDispatchIntent} from '../hooks/dashboard/useDispatchIntent'
export {useManageFavorite} from '../hooks/dashboard/useManageFavorite'
export {
  type NavigateToStudioResult,
  useNavigateToStudioDocument,
} from '../hooks/dashboard/useNavigateToStudioDocument'
export {useRecordDocumentHistoryEvent} from '../hooks/dashboard/useRecordDocumentHistoryEvent'
export {useStudioWorkspacesByProjectIdDataset} from '../hooks/dashboard/useStudioWorkspacesByProjectIdDataset'
export {useWindowTitle} from '../hooks/dashboard/useWindowTitle'
export {useDatasets} from '../hooks/datasets/useDatasets'
export {useApplyDocumentActions} from '../hooks/document/useApplyDocumentActions'
export {useDocument} from '../hooks/document/useDocument'
export {useDocumentEvent} from '../hooks/document/useDocumentEvent'
export {useDocumentPermissions} from '../hooks/document/useDocumentPermissions'
export {useDocumentSyncStatus} from '../hooks/document/useDocumentSyncStatus'
export {useEditDocument} from '../hooks/document/useEditDocument'
export {
  type DocumentsOptions,
  type DocumentsResponse,
  useDocuments,
} from '../hooks/documents/useDocuments'
export {
  type PaginatedDocumentsOptions,
  type PaginatedDocumentsResponse,
  usePaginatedDocuments,
} from '../hooks/paginatedDocuments/usePaginatedDocuments'
export {usePresence} from '../hooks/presence/usePresence'
export {
  useDocumentPreview,
  type useDocumentPreviewOptions,
  type useDocumentPreviewResults,
} from '../hooks/preview/useDocumentPreview'
export {
  useDocumentProjection,
  type useDocumentProjectionOptions,
  type useDocumentProjectionResults,
} from '../hooks/projection/useDocumentProjection'
export {useProject} from '../hooks/projects/useProject'
export {type ProjectWithoutMembers, useProjects} from '../hooks/projects/useProjects'
export {useQuery} from '../hooks/query/useQuery'
export {useActiveReleases} from '../hooks/releases/useActiveReleases'
export {usePerspective} from '../hooks/releases/usePerspective'
export {type UserResult, useUser} from '../hooks/users/useUser'
export {type UsersResult, useUsers} from '../hooks/users/useUsers'
export {REACT_SDK_VERSION} from '../version'
export {type DatasetsResponse, type SanityProjectMember} from '@sanity/client'
export type {Status as ComlinkStatus} from '@sanity/comlink'
export {type SanityDocument, type SortOrderingItem} from '@sanity/types'
