// @sanity/sdk-react/dist/index.js
import { jsx, jsxs, Fragment } from "react/jsx-runtime";
import { c } from "react-compiler-runtime";
import { ClientError, CorsOriginError } from "@sanity/client";
import { getAuthState, getNodeState, getIsInDashboardState, isStudioConfig, setAuthToken, AuthStateType, getLoginUrlState, observeOrganizationVerificationState, handleAuthCallback, logout, isProjectUserNotFoundClientError, getClientErrorApiDescription, getClientErrorApiBody, isImportError, getCorsErrorProjectId, createSanityInstance, agentGenerate, agentTransform, agentTranslate, agentPrompt, agentPatch, getTokenState, getCurrentUserState, getDashboardOrganizationId, getClientState, getOrCreateController, getOrCreateChannel, releaseChannel, isDatasetResource, isMediaLibraryResource, isCanvasResource, getFavoritesState, resolveFavoritesState, resolveDatasets, getDatasetsState, applyDocumentActions, resolveDocument, getDocumentState, subscribeDocumentEvents, getPermissionsState, getDocumentSyncStatus, editDocument, getQueryKey, parseQueryKey, getQueryState, resolveQuery, createGroqSearchFilter, getPresence, getProjectionState, resolveProjection, transformProjectionToPreview, PREVIEW_PROJECTION, resolveProject, getProjectState, resolveProjects, getProjectsState, getActiveReleasesState, getPerspectiveState, getUsersKey, parseUsersKey, getUsersState, resolveUsers, loadMoreUsers } from "@sanity/sdk";
export * from "@sanity/sdk";
import { createContext, useContext, useSyncExternalStore, useRef, useEffect, useState, Suspense, useMemo, StrictMode, useCallback, useInsertionEffect, useTransition } from "react";
import { ErrorBoundary } from "react-error-boundary";
import { SDK_CHANNEL_NAME, SDK_NODE_NAME } from "@sanity/message-protocol";
import { firstValueFrom, filter, identity, Observable, startWith, distinctUntilChanged, switchMap, EMPTY } from "rxjs";
import { initTelemetry, trackHookMounted } from "@sanity/sdk/_internal";
import { createRoot } from "react-dom/client";
const SanityInstanceContext = createContext(null), useSanityInstance = (config) => {
  const $ = c(3), instance = useContext(SanityInstanceContext);
  if (!instance)
    throw new Error(`SanityInstance context not found. ${config ? `Requested config: ${JSON.stringify(config, null, 2)}. ` : ""}Please ensure that your component is wrapped in a ResourceProvider or a SanityApp component.`);
  if (!config)
    return instance;
  let t0;
  $[0] !== config || $[1] !== instance ? (t0 = instance.match(config), $[0] = config, $[1] = instance, $[2] = t0) : t0 = $[2];
  const match = t0;
  if (!match)
    throw new Error(`Could not find a matching Sanity instance for the requested configuration: ${JSON.stringify(config, null, 2)}.
Please ensure there is a ResourceProvider component with a matching configuration in the component hierarchy.`);
  return match;
};
function createStateSourceHook(options) {
  const getState = typeof options == "function" ? options : options.getState, getConfig = "getConfig" in options ? options.getConfig : void 0, suspense = "shouldSuspend" in options && "suspender" in options ? options : void 0;
  function useHook(...t0) {
    const $ = c(5), params = t0;
    let t1;
    $[0] !== params ? (t1 = getConfig?.(...params), $[0] = params, $[1] = t1) : t1 = $[1];
    const instance = useSanityInstance(t1);
    if (suspense?.suspender && suspense?.shouldSuspend?.(instance, ...params))
      throw suspense.suspender(instance, ...params);
    let t2;
    $[2] !== instance || $[3] !== params ? (t2 = getState(instance, ...params), $[2] = instance, $[3] = params, $[4] = t2) : t2 = $[4];
    const state = t2;
    return useSyncExternalStore(state.subscribe, state.getCurrent);
  }
  return useHook;
}
const useAuthState = createStateSourceHook(getAuthState), useNodeState = createStateSourceHook({
  getState: getNodeState,
  shouldSuspend: (instance, nodeInput) => getNodeState(instance, nodeInput).getCurrent() === void 0,
  suspender: (instance, nodeInput) => firstValueFrom(getNodeState(instance, nodeInput).observable.pipe(filter(Boolean)))
});
function useWindowConnection(t0) {
  const $ = c(19), {
    name,
    connectTo,
    onMessage
  } = t0;
  let t1;
  $[0] !== connectTo || $[1] !== name ? (t1 = {
    name,
    connectTo
  }, $[0] = connectTo, $[1] = name, $[2] = t1) : t1 = $[2];
  const {
    node
  } = useNodeState(t1);
  let t2;
  $[3] === Symbol.for("react.memo_cache_sentinel") ? (t2 = [], $[3] = t2) : t2 = $[3];
  const messageUnsubscribers = useRef(t2), instance = useSanityInstance();
  let t3;
  $[4] !== node || $[5] !== onMessage ? (t3 = () => (onMessage && Object.entries(onMessage).forEach((t42) => {
    const [type, handler] = t42, messageUnsubscribe = node.on(type, handler);
    messageUnsubscribe && messageUnsubscribers.current.push(messageUnsubscribe);
  }), () => {
    messageUnsubscribers.current.forEach(_temp$9), messageUnsubscribers.current = [];
  }), $[4] = node, $[5] = onMessage, $[6] = t3) : t3 = $[6];
  let t4;
  $[7] !== instance || $[8] !== name || $[9] !== node || $[10] !== onMessage ? (t4 = [instance, name, onMessage, node], $[7] = instance, $[8] = name, $[9] = node, $[10] = onMessage, $[11] = t4) : t4 = $[11], useEffect(t3, t4);
  let t5;
  $[12] !== node ? (t5 = (type_0, data) => {
    node.post(type_0, data);
  }, $[12] = node, $[13] = t5) : t5 = $[13];
  const sendMessage = t5;
  let t6;
  $[14] !== node ? (t6 = (type_1, data_0, fetchOptions) => node.fetch(type_1, data_0, fetchOptions ?? {}), $[14] = node, $[15] = t6) : t6 = $[15];
  const fetch = t6;
  let t7;
  return $[16] !== fetch || $[17] !== sendMessage ? (t7 = {
    sendMessage,
    fetch
  }, $[16] = fetch, $[17] = sendMessage, $[18] = t7) : t7 = $[18], t7;
}
function _temp$9(unsubscribe) {
  return unsubscribe();
}
const DEFAULT_RESPONSE_TIMEOUT = 1e4;
function DashboardTokenRefresh(t0) {
  const $ = c(14), {
    children
  } = t0, instance = useSanityInstance(), isTokenRefreshInProgress = useRef(!1), timeoutRef = useRef(null), processed401ErrorRef = useRef(null), authState = useAuthState();
  let t1;
  $[0] === Symbol.for("react.memo_cache_sentinel") ? (t1 = () => {
    timeoutRef.current && (clearTimeout(timeoutRef.current), timeoutRef.current = null);
  }, $[0] = t1) : t1 = $[0];
  const clearRefreshTimeout = t1;
  let t2;
  $[1] === Symbol.for("react.memo_cache_sentinel") ? (t2 = {
    name: SDK_NODE_NAME,
    connectTo: SDK_CHANNEL_NAME
  }, $[1] = t2) : t2 = $[1];
  const windowConnection = useWindowConnection(t2);
  let t3;
  $[2] !== instance || $[3] !== windowConnection ? (t3 = async () => {
    if (!isTokenRefreshInProgress.current) {
      isTokenRefreshInProgress.current = !0, clearRefreshTimeout(), timeoutRef.current = setTimeout(() => {
        isTokenRefreshInProgress.current && (isTokenRefreshInProgress.current = !1), timeoutRef.current = null;
      }, DEFAULT_RESPONSE_TIMEOUT);
      try {
        const res = await windowConnection.fetch("dashboard/v1/auth/tokens/create");
        if (clearRefreshTimeout(), res.token) {
          setAuthToken(instance, res.token);
          const errorContainer = document.getElementById("__sanityError");
          errorContainer && Array.from(errorContainer.getElementsByTagName("div")).some(_temp$8) && errorContainer.remove();
        }
        isTokenRefreshInProgress.current = !1;
      } catch {
        isTokenRefreshInProgress.current = !1, clearRefreshTimeout();
      }
    }
  }, $[2] = instance, $[3] = windowConnection, $[4] = t3) : t3 = $[4];
  const requestNewToken = t3;
  let t4, t5;
  $[5] === Symbol.for("react.memo_cache_sentinel") ? (t4 = () => () => {
    clearRefreshTimeout();
  }, t5 = [clearRefreshTimeout], $[5] = t4, $[6] = t5) : (t4 = $[5], t5 = $[6]), useEffect(t4, t5);
  let t6;
  $[7] !== authState.error || $[8] !== authState.type || $[9] !== requestNewToken ? (t6 = () => {
    const has401Error = authState.type === AuthStateType.ERROR && authState.error && authState.error?.statusCode === 401 && !isTokenRefreshInProgress.current && processed401ErrorRef.current !== authState.error, isLoggedOut = authState.type === AuthStateType.LOGGED_OUT && !isTokenRefreshInProgress.current;
    has401Error || isLoggedOut ? (processed401ErrorRef.current = authState.type === AuthStateType.ERROR ? authState.error : void 0, requestNewToken()) : (authState.type !== AuthStateType.ERROR || processed401ErrorRef.current !== (authState.type === AuthStateType.ERROR ? authState.error : void 0)) && (processed401ErrorRef.current = null);
  }, $[7] = authState.error, $[8] = authState.type, $[9] = requestNewToken, $[10] = t6) : t6 = $[10];
  let t7;
  return $[11] !== authState || $[12] !== requestNewToken ? (t7 = [authState, requestNewToken], $[11] = authState, $[12] = requestNewToken, $[13] = t7) : t7 = $[13], useEffect(t6, t7), children;
}
function _temp$8(div) {
  return div.textContent?.includes("Uncaught error: Unauthorized - A valid session is required for this endpoint");
}
const ComlinkTokenRefreshProvider = (t0) => {
  const $ = c(2), {
    children
  } = t0, instance = useSanityInstance();
  let t1;
  t1 = getIsInDashboardState(instance).getCurrent();
  const isInDashboard = t1, isStudio = isStudioConfig(instance.config);
  if (isInDashboard && !isStudio) {
    let t2;
    return $[0] !== children ? (t2 = /* @__PURE__ */ jsx(DashboardTokenRefresh, { children }), $[0] = children, $[1] = t2) : t2 = $[1], t2;
  }
  return children;
};
function useLoginUrl() {
  const $ = c(2), instance = useSanityInstance();
  let t0, t1;
  $[0] !== instance ? (t1 = getLoginUrlState(instance), $[0] = instance, $[1] = t1) : t1 = $[1], t0 = t1;
  const {
    subscribe,
    getCurrent
  } = t0;
  return useSyncExternalStore(subscribe, getCurrent);
}
function useVerifyOrgProjects(t0, projectIds) {
  const $ = c(5), disabled = t0 === void 0 ? !1 : t0, instance = useSanityInstance(), [error, setError] = useState(null), isInactive = disabled || !projectIds || projectIds.length === 0, [prevInactive, setPrevInactive] = useState(isInactive);
  prevInactive !== isInactive && (setPrevInactive(isInactive), isInactive && setError(null));
  let t1, t2;
  return $[0] !== instance || $[1] !== isInactive || $[2] !== projectIds ? (t1 = () => {
    if (isInactive)
      return;
    const subscription = observeOrganizationVerificationState(instance, projectIds).subscribe((result) => {
      setError(result.error);
    });
    return () => {
      subscription.unsubscribe();
    };
  }, t2 = [instance, isInactive, projectIds], $[0] = instance, $[1] = isInactive, $[2] = projectIds, $[3] = t1, $[4] = t2) : (t1 = $[3], t2 = $[4]), useEffect(t1, t2), error;
}
const CHUNK_RELOAD_STORAGE_KEY = "__sanity_sdk_chunk_reload_attempted";
function readChunkReloadFlag() {
  try {
    return typeof window > "u" || typeof window.sessionStorage > "u" ? !1 : window.sessionStorage.getItem(CHUNK_RELOAD_STORAGE_KEY) !== null;
  } catch {
    return !1;
  }
}
function setChunkReloadFlag() {
  try {
    if (typeof window > "u" || typeof window.sessionStorage > "u") return;
    window.sessionStorage.setItem(CHUNK_RELOAD_STORAGE_KEY, "1");
  } catch {
  }
}
function clearChunkReloadFlag() {
  try {
    if (typeof window > "u" || typeof window.sessionStorage > "u") return;
    window.sessionStorage.removeItem(CHUNK_RELOAD_STORAGE_KEY);
  } catch {
  }
}
const FONT_SANS_SERIF = "-apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Helvetica, Arial, system-ui, sans-serif", FONT_MONOSPACE = "-apple-system-ui-monospace, 'SF Mono', Menlo, Monaco, Consolas, monospace", styles = {
  container: {
    padding: "28px",
    fontFamily: FONT_SANS_SERIF,
    display: "flex",
    flexDirection: "column",
    gap: "21px",
    fontSize: "14px"
  },
  heading: {
    margin: 0,
    fontSize: "28px",
    fontWeight: 700
  },
  paragraph: {
    margin: 0
  },
  link: {
    appearance: "none",
    background: "transparent",
    border: 0,
    padding: 0,
    font: "inherit",
    textDecoration: "underline",
    cursor: "pointer"
  },
  code: {
    fontFamily: FONT_MONOSPACE
  }
};
function Error$1(t0) {
  const $ = c(13), {
    heading,
    description,
    code,
    cta
  } = t0;
  let t1;
  $[0] !== heading ? (t1 = /* @__PURE__ */ jsx("h1", { style: styles.heading, children: heading }), $[0] = heading, $[1] = t1) : t1 = $[1];
  let t2;
  $[2] !== description ? (t2 = description && /* @__PURE__ */ jsx("p", { style: styles.paragraph, dangerouslySetInnerHTML: {
    __html: description
  } }), $[2] = description, $[3] = t2) : t2 = $[3];
  let t3;
  $[4] !== code ? (t3 = code && /* @__PURE__ */ jsx("code", { style: styles.code, children: code }), $[4] = code, $[5] = t3) : t3 = $[5];
  let t4;
  $[6] !== cta ? (t4 = cta && (cta.href || cta.onClick) && /* @__PURE__ */ jsx("p", { style: styles.paragraph, children: cta.href ? /* @__PURE__ */ jsx("a", { style: styles.link, href: cta.href, target: "_blank", rel: "noopener noreferrer", children: cta.text }) : /* @__PURE__ */ jsx("button", { style: styles.link, onClick: cta.onClick, children: cta.text }) }), $[6] = cta, $[7] = t4) : t4 = $[7];
  let t5;
  return $[8] !== t1 || $[9] !== t2 || $[10] !== t3 || $[11] !== t4 ? (t5 = /* @__PURE__ */ jsxs("div", { style: styles.container, children: [
    t1,
    t2,
    t3,
    t4
  ] }), $[8] = t1, $[9] = t2, $[10] = t3, $[11] = t4, $[12] = t5) : t5 = $[12], t5;
}
function reload() {
  try {
    window.location.reload();
  } catch {
  }
}
function ChunkLoadError(_props) {
  const $ = c(4);
  let t0;
  $[0] === Symbol.for("react.memo_cache_sentinel") ? (t0 = readChunkReloadFlag(), $[0] = t0) : t0 = $[0];
  const alreadyAttempted = t0;
  let t1, t2;
  if ($[1] === Symbol.for("react.memo_cache_sentinel") ? (t1 = () => {
    alreadyAttempted || (setChunkReloadFlag(), reload());
  }, t2 = [alreadyAttempted], $[1] = t1, $[2] = t2) : (t1 = $[1], t2 = $[2]), useEffect(t1, t2), !alreadyAttempted)
    return null;
  let t3;
  return $[3] === Symbol.for("react.memo_cache_sentinel") ? (t3 = /* @__PURE__ */ jsx(Error$1, { heading: "A new version is available", description: "The page tried to load an asset that no longer exists. Reload to continue with the latest version.", cta: {
    text: "Reload page",
    onClick: _temp$7
  } }), $[3] = t3) : t3 = $[3], t3;
}
function _temp$7() {
  clearChunkReloadFlag(), reload();
}
function CorsErrorComponent(t0) {
  const $ = c(6), {
    projectId,
    error
  } = t0, origin = window.location.origin;
  let t1;
  const t2 = `https://sanity.io/manage/project/${projectId}/api`;
  let t3;
  if ($[0] !== t2) {
    const url = new URL(t2);
    url.searchParams.set("cors", "add"), url.searchParams.set("origin", origin), url.searchParams.set("credentials", "include"), t3 = url.toString(), $[0] = t2, $[1] = t3;
  } else
    t3 = $[1];
  t1 = t3;
  const corsUrl = t1;
  let t4;
  return $[2] !== corsUrl || $[3] !== error?.message || $[4] !== projectId ? (t4 = /* @__PURE__ */ jsx(Error$1, { heading: "Before you continue\u2026", ...projectId ? {
    description: "To access your content, you need to <strong>add the following URL as a CORS origin</strong> to your Sanity project.",
    code: origin,
    cta: {
      text: "Manage CORS configuration",
      href: corsUrl
    }
  } : {
    description: error?.message
  } }), $[2] = corsUrl, $[3] = error?.message, $[4] = projectId, $[5] = t4) : t4 = $[5], t4;
}
function isInIframe() {
  return typeof window < "u" && window.self !== window.top;
}
function isLocalUrl(window2) {
  const url = typeof window2 < "u" ? window2.location.href : "";
  return url.startsWith("http://localhost") || url.startsWith("https://localhost") || url.startsWith("http://127.0.0.1") || url.startsWith("https://127.0.0.1");
}
class AuthError extends Error {
  constructor(error) {
    typeof error == "object" && error && "message" in error && typeof error.message == "string" ? super(error.message) : super(), this.cause = error;
  }
}
class ConfigurationError extends Error {
  constructor(error) {
    typeof error == "object" && error && "message" in error && typeof error.message == "string" ? super(error.message) : super(), this.cause = error;
  }
}
function createCallbackHook(callback) {
  function useHook() {
    const $ = c(2), instance = useSanityInstance();
    let t0;
    return $[0] !== instance ? (t0 = (...t1) => callback(instance, ...t1), $[0] = instance, $[1] = t0) : t0 = $[1], t0;
  }
  return useHook;
}
const useHandleAuthCallback = createCallbackHook(handleAuthCallback);
function LoginCallback() {
  const $ = c(3), handleAuthCallback2 = useHandleAuthCallback();
  let t0, t1;
  return $[0] !== handleAuthCallback2 ? (t0 = () => {
    const url = new URL(location.href);
    handleAuthCallback2(url.toString()).then(_temp$6);
  }, t1 = [handleAuthCallback2], $[0] = handleAuthCallback2, $[1] = t0, $[2] = t1) : (t0 = $[1], t1 = $[2]), useEffect(t0, t1), null;
}
function _temp$6(replacementLocation) {
  replacementLocation && history.replaceState(null, "", replacementLocation);
}
const useLogOut = createCallbackHook(logout);
function DashboardAccessRequest(t0) {
  const $ = c(5), {
    projectId
  } = t0;
  let t1;
  $[0] === Symbol.for("react.memo_cache_sentinel") ? (t1 = {
    name: SDK_NODE_NAME,
    connectTo: SDK_CHANNEL_NAME
  }, $[0] = t1) : t1 = $[0];
  const {
    fetch
  } = useWindowConnection(t1);
  let t2, t3;
  return $[1] !== fetch || $[2] !== projectId ? (t2 = () => {
    fetch("dashboard/v1/auth/access/request", {
      resourceType: "project",
      resourceId: projectId
    });
  }, t3 = [fetch, projectId], $[1] = fetch, $[2] = projectId, $[3] = t2, $[4] = t3) : (t2 = $[3], t3 = $[4]), useEffect(t2, t3), null;
}
function LoginError(t0) {
  const $ = c(30), {
    error,
    resetErrorBoundary
  } = t0;
  if (!(error instanceof AuthError || error instanceof ConfigurationError || error instanceof ClientError))
    throw error;
  const logout2 = useLogOut(), authState = useAuthState(), instance = useSanityInstance(), {
    config: t1
  } = instance, {
    projectId
  } = t1, clientError = error instanceof ClientError ? error : error instanceof AuthError && error.cause instanceof ClientError ? error.cause : null;
  let t2;
  $[0] !== instance ? (t2 = getIsInDashboardState(instance).getCurrent(), $[0] = instance, $[1] = t2) : t2 = $[1];
  const isInDashboard = t2;
  let t3;
  $[2] !== clientError ? (t3 = !!clientError && clientError.statusCode === 401 && isProjectUserNotFoundClientError(clientError), $[2] = clientError, $[3] = t3) : t3 = $[3];
  const isProjectUserNotFound = t3, dashboardAccessProjectId = isProjectUserNotFound && projectId && isInDashboard ? projectId : null;
  let t4;
  $[4] !== logout2 || $[5] !== resetErrorBoundary ? (t4 = async () => {
    await logout2(), resetErrorBoundary();
  }, $[4] = logout2, $[5] = resetErrorBoundary, $[6] = t4) : t4 = $[6];
  const handleRetry = t4;
  let t5, message = "Please try again or contact support if the problem persists.", retry = !0;
  if (clientError) {
    if (clientError.statusCode === 401)
      if (isProjectUserNotFound) {
        let t62;
        $[7] !== clientError ? (t62 = getClientErrorApiDescription(clientError), $[7] = clientError, $[8] = t62) : t62 = $[8];
        const description = t62;
        description && (message = description), retry = !1;
      } else
        isInDashboard || (message = "Signing you out and returning to login...", retry = !0);
    else if (clientError.statusCode === 404) {
      const errorMessage = getClientErrorApiBody(clientError)?.message || "";
      message = errorMessage.startsWith("Session with sid") && errorMessage.endsWith("not found") ? "The session ID is invalid or expired." : "The login link is invalid or expired. Please try again.", retry = !0;
    }
  }
  authState.type !== AuthStateType.ERROR && error instanceof ConfigurationError && (message = error.message, retry = !0);
  let t6;
  $[9] !== message || $[10] !== retry ? (t6 = {
    authErrorMessage: message,
    showRetryCta: retry
  }, $[9] = message, $[10] = retry, $[11] = t6) : t6 = $[11], t5 = t6;
  const {
    authErrorMessage,
    showRetryCta
  } = t5, hasAutoLoggedOutRef = useRef(!1);
  let t7, t8;
  $[12] !== clientError || $[13] !== handleRetry || $[14] !== isInDashboard || $[15] !== isProjectUserNotFound ? (t7 = () => {
    clientError && clientError.statusCode === 401 && !isProjectUserNotFound && !isInDashboard && !hasAutoLoggedOutRef.current && (hasAutoLoggedOutRef.current = !0, handleRetry());
  }, t8 = [clientError, handleRetry, isInDashboard, isProjectUserNotFound], $[12] = clientError, $[13] = handleRetry, $[14] = isInDashboard, $[15] = isProjectUserNotFound, $[16] = t7, $[17] = t8) : (t7 = $[16], t8 = $[17]), useEffect(t7, t8);
  let t9;
  $[18] !== dashboardAccessProjectId ? (t9 = dashboardAccessProjectId && /* @__PURE__ */ jsx(Suspense, { fallback: null, children: /* @__PURE__ */ jsx(DashboardAccessRequest, { projectId: dashboardAccessProjectId }) }), $[18] = dashboardAccessProjectId, $[19] = t9) : t9 = $[19];
  const t10 = error instanceof ConfigurationError ? "Configuration Error" : "Authentication Error";
  let t11;
  $[20] !== handleRetry || $[21] !== showRetryCta ? (t11 = showRetryCta ? {
    text: "Retry",
    onClick: handleRetry
  } : void 0, $[20] = handleRetry, $[21] = showRetryCta, $[22] = t11) : t11 = $[22];
  let t12;
  $[23] !== authErrorMessage || $[24] !== t10 || $[25] !== t11 ? (t12 = /* @__PURE__ */ jsx(Error$1, { heading: t10, description: authErrorMessage, cta: t11 }), $[23] = authErrorMessage, $[24] = t10, $[25] = t11, $[26] = t12) : t12 = $[26];
  let t13;
  return $[27] !== t12 || $[28] !== t9 ? (t13 = /* @__PURE__ */ jsxs(Fragment, { children: [
    t9,
    t12
  ] }), $[27] = t12, $[28] = t9, $[29] = t13) : t13 = $[29], t13;
}
if (isInIframe() && !document.querySelector("[data-sanity-core]")) {
  const parsedUrl = new URL(window.location.href), mode = new URLSearchParams(parsedUrl.hash.slice(1)).get("mode"), script = document.createElement("script");
  script.src = mode === "core-ui--staging" ? "https://core.sanity-cdn.work/bridge.js" : "https://core.sanity-cdn.com/bridge.js", script.type = "module", script.async = !0, document.head.appendChild(script);
}
function AuthBoundary(t0) {
  const $ = c(10);
  let props, t1;
  $[0] !== t0 ? ({
    LoginErrorComponent: t1,
    ...props
  } = t0, $[0] = t0, $[1] = props, $[2] = t1) : (props = $[1], t1 = $[2]);
  const LoginErrorComponent = t1 === void 0 ? LoginError : t1;
  let t2, t3;
  $[3] !== LoginErrorComponent ? (t3 = function(fallbackProps) {
    return isImportError(fallbackProps.error) ? /* @__PURE__ */ jsx(ChunkLoadError, { ...fallbackProps }) : fallbackProps.error instanceof CorsOriginError ? /* @__PURE__ */ jsx(CorsErrorComponent, { ...fallbackProps, projectId: getCorsErrorProjectId(fallbackProps.error) }) : /* @__PURE__ */ jsx(LoginErrorComponent, { ...fallbackProps });
  }, $[3] = LoginErrorComponent, $[4] = t3) : t3 = $[4], t2 = t3;
  const FallbackComponent = t2;
  let t4;
  $[5] !== props ? (t4 = /* @__PURE__ */ jsx(AuthSwitch, { ...props }), $[5] = props, $[6] = t4) : t4 = $[6];
  let t5;
  return $[7] !== FallbackComponent || $[8] !== t4 ? (t5 = /* @__PURE__ */ jsx(ComlinkTokenRefreshProvider, { children: /* @__PURE__ */ jsx(ErrorBoundary, { FallbackComponent, children: t4 }) }), $[7] = FallbackComponent, $[8] = t4, $[9] = t5) : t5 = $[9], t5;
}
function AuthSwitch(t0) {
  const $ = c(16);
  let children, projectIds, props, t1, t2;
  $[0] !== t0 ? ({
    CallbackComponent: t1,
    children,
    verifyOrganization: t2,
    projectIds,
    ...props
  } = t0, $[0] = t0, $[1] = children, $[2] = projectIds, $[3] = props, $[4] = t1, $[5] = t2) : (children = $[1], projectIds = $[2], props = $[3], t1 = $[4], t2 = $[5]);
  const CallbackComponent = t1 === void 0 ? LoginCallback : t1, verifyOrganization = t2 === void 0 ? !0 : t2, authState = useAuthState(), instance = useSanityInstance();
  let t3;
  $[6] !== instance.config ? (t3 = isStudioConfig(instance.config), $[6] = instance.config, $[7] = t3) : t3 = $[7];
  const isStudio = t3, disableVerifyOrg = !verifyOrganization || isStudio || authState.type !== AuthStateType.LOGGED_IN, orgError = useVerifyOrgProjects(disableVerifyOrg, projectIds), isLoggedOut = authState.type === AuthStateType.LOGGED_OUT && !authState.isDestroyingSession, loginUrl = useLoginUrl();
  let t4, t5;
  if ($[8] !== isLoggedOut || $[9] !== isStudio || $[10] !== loginUrl ? (t4 = () => {
    isLoggedOut && !isInIframe() && !isStudio && (window.location.href = loginUrl);
  }, t5 = [isLoggedOut, loginUrl, isStudio], $[8] = isLoggedOut, $[9] = isStudio, $[10] = loginUrl, $[11] = t4, $[12] = t5) : (t4 = $[11], t5 = $[12]), useEffect(t4, t5), verifyOrganization && orgError)
    throw new ConfigurationError({
      message: orgError
    });
  switch (authState.type) {
    case AuthStateType.ERROR:
      throw new AuthError(authState.error);
    case AuthStateType.LOGGING_IN: {
      let t6;
      return $[13] !== CallbackComponent || $[14] !== props ? (t6 = /* @__PURE__ */ jsx(CallbackComponent, { ...props }), $[13] = CallbackComponent, $[14] = props, $[15] = t6) : t6 = $[15], t6;
    }
    case AuthStateType.LOGGED_IN:
      return children;
    case AuthStateType.LOGGED_OUT:
      return null;
    default:
      throw new Error(`Invalid auth state: ${authState.type}`);
  }
}
const SDKStudioContext = createContext(null);
SDKStudioContext.displayName = "SDKStudioContext";
function SanityInstanceProvider(t0) {
  const $ = c(6), {
    instance,
    fallback,
    children
  } = t0;
  let t1;
  $[0] !== children || $[1] !== fallback ? (t1 = /* @__PURE__ */ jsx(Suspense, { fallback, children }), $[0] = children, $[1] = fallback, $[2] = t1) : t1 = $[2];
  let t2;
  return $[3] !== instance || $[4] !== t1 ? (t2 = /* @__PURE__ */ jsx(SanityInstanceContext.Provider, { value: instance, children: t1 }), $[3] = instance, $[4] = t1, $[5] = t2) : t2 = $[5], t2;
}
const DEFAULT_FALLBACK = /* @__PURE__ */ jsx(Fragment, { children: "Warning: No fallback provided. Please supply a fallback prop to ensure proper Suspense handling." });
function ResourceProvider({
  children,
  fallback,
  ...config
}) {
  const parent = useContext(SanityInstanceContext), instance = useMemo(() => parent ? parent.createChild(config) : createSanityInstance(config), [config, parent]), projectId = config.projectId ?? "";
  useMemo(() => {
    projectId && !parent && initTelemetry(instance, projectId);
  }, [instance, projectId, parent]);
  const disposal = useRef(null);
  return useEffect(() => (disposal.current !== null && instance === disposal.current.instance && (clearTimeout(disposal.current.timeoutId), disposal.current = null), () => {
    disposal.current = {
      instance,
      timeoutId: setTimeout(() => {
        instance.isDisposed() || instance.dispose();
      }, 0)
    };
  }), [instance]), /* @__PURE__ */ jsx(SanityInstanceProvider, { instance, fallback: fallback ?? DEFAULT_FALLBACK, children });
}
const ResourcesContext = createContext({});
function ResetChunkReloadFlagOnMount() {
  const $ = c(1);
  let t0;
  return $[0] === Symbol.for("react.memo_cache_sentinel") ? (t0 = [], $[0] = t0) : t0 = $[0], useEffect(_temp$5, t0), null;
}
function _temp$5() {
  clearChunkReloadFlag();
}
function SDKProvider(t0) {
  const $ = c(20);
  let children, config, fallback, props;
  $[0] !== t0 ? ({
    children,
    config,
    fallback,
    ...props
  } = t0, $[0] = t0, $[1] = children, $[2] = config, $[3] = fallback, $[4] = props) : (children = $[1], config = $[2], fallback = $[3], props = $[4]);
  let t1;
  $[5] !== config ? (t1 = Array.isArray(config) ? config : [config], $[5] = config, $[6] = t1) : t1 = $[6];
  let configs, t2;
  $[7] !== t1 ? (configs = t1.slice().reverse(), t2 = configs.map(_temp2$1).filter(_temp3$1), $[7] = t1, $[8] = configs, $[9] = t2) : (configs = $[8], t2 = $[9]);
  const projectIds = t2;
  let t3, t4;
  $[10] !== props.resources ? (t4 = props.resources ?? {}, $[10] = props.resources, $[11] = t4) : t4 = $[11], t3 = t4;
  const resourcesValue = t3;
  let t5;
  if ($[12] !== children || $[13] !== configs || $[14] !== fallback || $[15] !== projectIds || $[16] !== props || $[17] !== resourcesValue) {
    const createNestedProviders = (index) => index >= configs.length ? /* @__PURE__ */ jsx(AuthBoundary, { ...props, projectIds, children: /* @__PURE__ */ jsx(ResourcesContext.Provider, { value: resourcesValue, children }) }) : /* @__PURE__ */ jsx(ResourceProvider, { ...configs[index], fallback, children: createNestedProviders(index + 1) });
    let t6;
    $[19] === Symbol.for("react.memo_cache_sentinel") ? (t6 = /* @__PURE__ */ jsx(ResetChunkReloadFlagOnMount, {}), $[19] = t6) : t6 = $[19], t5 = /* @__PURE__ */ jsxs(ErrorBoundary, { FallbackComponent: ChunkAwareFallback, children: [
      t6,
      createNestedProviders(0)
    ] }), $[12] = children, $[13] = configs, $[14] = fallback, $[15] = projectIds, $[16] = props, $[17] = resourcesValue, $[18] = t5;
  } else
    t5 = $[18];
  return t5;
}
function _temp3$1(id) {
  return !!id;
}
function _temp2$1(c2) {
  return c2.projectId;
}
function ChunkAwareFallback(fallbackProps) {
  const $ = c(2);
  if (isImportError(fallbackProps.error)) {
    let t0;
    return $[0] !== fallbackProps ? (t0 = /* @__PURE__ */ jsx(ChunkLoadError, { ...fallbackProps }), $[0] = fallbackProps, $[1] = t0) : t0 = $[1], t0;
  }
  throw fallbackProps.error;
}
const REDIRECT_URL = "https://sanity.io/welcome";
function deriveConfigFromWorkspace(workspace) {
  return {
    projectId: workspace.projectId,
    dataset: workspace.dataset,
    studio: {
      authenticated: workspace.authenticated,
      auth: workspace.auth.token ? {
        token: workspace.auth.token
      } : void 0
    }
  };
}
function SanityApp(t0) {
  const $ = c(18);
  let children, configProp, fallback, props;
  $[0] !== t0 ? ({
    children,
    fallback,
    config: configProp,
    ...props
  } = t0, $[0] = t0, $[1] = children, $[2] = configProp, $[3] = fallback, $[4] = props) : (children = $[1], configProp = $[2], fallback = $[3], props = $[4]);
  const studioWorkspace = useContext(SDKStudioContext);
  let t1;
  bb0: {
    if (configProp) {
      t1 = configProp;
      break bb0;
    }
    if (studioWorkspace) {
      let t23;
      $[5] !== studioWorkspace ? (t23 = deriveConfigFromWorkspace(studioWorkspace), $[5] = studioWorkspace, $[6] = t23) : t23 = $[6], t1 = t23;
      break bb0;
    }
    let t22;
    $[7] === Symbol.for("react.memo_cache_sentinel") ? (t22 = [], $[7] = t22) : t22 = $[7], t1 = t22;
  }
  const resolvedConfig = t1;
  let t2, t3;
  $[8] !== configProp || $[9] !== resolvedConfig || $[10] !== studioWorkspace ? (t2 = () => {
    let timeout;
    const primaryConfig = Array.isArray(resolvedConfig) ? resolvedConfig[0] : resolvedConfig, shouldRedirectWithoutConfig = configProp === void 0 && !studioWorkspace && !primaryConfig;
    return !isInIframe() && !isLocalUrl(window) && (shouldRedirectWithoutConfig || primaryConfig && !isStudioConfig(primaryConfig)) && (timeout = setTimeout(_temp$4, 1e3)), () => clearTimeout(timeout);
  }, t3 = [configProp, resolvedConfig, studioWorkspace], $[8] = configProp, $[9] = resolvedConfig, $[10] = studioWorkspace, $[11] = t2, $[12] = t3) : (t2 = $[11], t3 = $[12]), useEffect(t2, t3);
  let t4;
  return $[13] !== children || $[14] !== fallback || $[15] !== props || $[16] !== resolvedConfig ? (t4 = /* @__PURE__ */ jsx(SDKProvider, { ...props, fallback, config: resolvedConfig, children }), $[13] = children, $[14] = fallback, $[15] = props, $[16] = resolvedConfig, $[17] = t4) : t4 = $[17], t4;
}
function _temp$4() {
  console.warn("Redirecting to core", REDIRECT_URL), window.location.replace(REDIRECT_URL);
}
function renderSanityApp(rootElement, namedSources, options, children) {
  if (!rootElement)
    throw new Error("Missing root element to mount application into");
  const {
    reactStrictMode = !1
  } = options, root = createRoot(rootElement), config = Object.values(namedSources);
  return root.render(reactStrictMode ? /* @__PURE__ */ jsx(StrictMode, { children: /* @__PURE__ */ jsx(SanityApp, { config, fallback: /* @__PURE__ */ jsx("div", { children: "Loading..." }), children }) }) : /* @__PURE__ */ jsx(SanityApp, { config, fallback: /* @__PURE__ */ jsx("div", { children: "Loading..." }), children })), () => root.unmount();
}
const useAgentGenerate = createCallbackHook(agentGenerate), useAgentTransform = createCallbackHook(agentTransform), useAgentTranslate = createCallbackHook(agentTranslate);
function promptAdapter(instance, options) {
  return firstValueFrom(agentPrompt(instance, options));
}
const useAgentPrompt = createCallbackHook(promptAdapter);
function patchAdapter(instance, options) {
  return firstValueFrom(agentPatch(instance, options));
}
const useAgentPatch = createCallbackHook(patchAdapter);
function useAgentResourceContext(options) {
  const $ = c(9), {
    projectId,
    dataset,
    documentId
  } = options;
  let t0;
  $[0] === Symbol.for("react.memo_cache_sentinel") ? (t0 = {
    name: SDK_NODE_NAME,
    connectTo: SDK_CHANNEL_NAME
  }, $[0] = t0) : t0 = $[0];
  const {
    sendMessage
  } = useWindowConnection(t0), lastContextRef = useRef(null);
  let t1;
  $[1] !== dataset || $[2] !== documentId || $[3] !== projectId || $[4] !== sendMessage ? (t1 = () => {
    if (!projectId || !dataset) {
      console.warn("[useAgentResourceContext] projectId and dataset are required", {
        projectId,
        dataset
      });
      return;
    }
    const contextKey = `${projectId}:${dataset}:${documentId || ""}`;
    if (lastContextRef.current !== contextKey)
      try {
        const message = {
          type: "dashboard/v1/events/agent/resource/update",
          data: {
            projectId,
            dataset,
            documentId
          }
        };
        sendMessage(message.type, message.data), lastContextRef.current = contextKey;
      } catch (t22) {
        console.error("[useAgentResourceContext] Failed to update context:", t22);
      }
  }, $[1] = dataset, $[2] = documentId, $[3] = projectId, $[4] = sendMessage, $[5] = t1) : t1 = $[5];
  const updateContext = t1;
  let t2, t3;
  $[6] !== updateContext ? (t2 = () => {
    updateContext();
  }, t3 = [updateContext], $[6] = updateContext, $[7] = t2, $[8] = t3) : (t2 = $[7], t3 = $[8]), useEffect(t2, t3);
}
const useAuthToken = createStateSourceHook(getTokenState), useCurrentUser = createStateSourceHook(getCurrentUserState);
function useDashboardOrganizationId() {
  const $ = c(2), instance = useSanityInstance();
  let t0, t1;
  $[0] !== instance ? (t1 = getDashboardOrganizationId(instance), $[0] = instance, $[1] = t1) : t1 = $[1], t0 = t1;
  const {
    subscribe,
    getCurrent
  } = t0;
  return useSyncExternalStore(subscribe, getCurrent);
}
const useClient = createStateSourceHook({
  getState: (instance, options) => {
    if (!options || typeof options != "object")
      throw new Error('useClient() requires a configuration object with at least an "apiVersion" property. Example: useClient({ apiVersion: "2024-11-12" })');
    return getClientState(instance, options);
  },
  getConfig: (options) => options
});
function useFrameConnection(options) {
  const $ = c(12), {
    onMessage,
    targetOrigin,
    name,
    connectTo,
    heartbeat,
    onStatus
  } = options, instance = useSanityInstance(), controllerRef = useRef(null), channelRef = useRef(null);
  let t0, t1;
  $[0] !== connectTo || $[1] !== heartbeat || $[2] !== instance || $[3] !== name || $[4] !== onMessage || $[5] !== onStatus || $[6] !== targetOrigin ? (t0 = () => {
    const controller = getOrCreateController(instance, targetOrigin), channel = getOrCreateChannel(instance, {
      name,
      connectTo,
      heartbeat
    });
    controllerRef.current = controller, channelRef.current = channel, channel.onStatus((event) => {
      onStatus?.(event.status);
    });
    const messageUnsubscribers = [];
    return onMessage && Object.entries(onMessage).forEach((t22) => {
      const [type, handler] = t22, unsubscribe = channel.on(type, handler);
      messageUnsubscribers.push(unsubscribe);
    }), () => {
      messageUnsubscribers.forEach(_temp$3), releaseChannel(instance, name), channelRef.current = null, controllerRef.current = null;
    };
  }, t1 = [targetOrigin, name, connectTo, heartbeat, onMessage, instance, onStatus], $[0] = connectTo, $[1] = heartbeat, $[2] = instance, $[3] = name, $[4] = onMessage, $[5] = onStatus, $[6] = targetOrigin, $[7] = t0, $[8] = t1) : (t0 = $[7], t1 = $[8]), useEffect(t0, t1);
  let t2;
  $[9] === Symbol.for("react.memo_cache_sentinel") ? (t2 = (frameWindow) => {
    const removeTarget = controllerRef.current?.addTarget(frameWindow);
    return () => {
      removeTarget?.();
    };
  }, $[9] = t2) : t2 = $[9];
  const connect = t2;
  let t3;
  $[10] === Symbol.for("react.memo_cache_sentinel") ? (t3 = (type_0, data) => {
    channelRef.current?.post(type_0, data);
  }, $[10] = t3) : t3 = $[10];
  const sendMessage = t3;
  let t4;
  return $[11] === Symbol.for("react.memo_cache_sentinel") ? (t4 = {
    connect,
    sendMessage
  }, $[11] = t4) : t4 = $[11], t4;
}
function _temp$3(unsub) {
  return unsub();
}
function useDashboardNavigate(navigateFn) {
  const $ = c(2);
  let t0;
  $[0] !== navigateFn ? (t0 = {
    name: SDK_NODE_NAME,
    connectTo: SDK_CHANNEL_NAME,
    onMessage: {
      "dashboard/v1/history/change-path": (data) => {
        navigateFn(data);
      }
    }
  }, $[0] = navigateFn, $[1] = t0) : t0 = $[1], useWindowConnection(t0);
}
function normalizeResourceOptions(options, resources) {
  const {
    resourceName,
    sourceName,
    source,
    ...rest
  } = options, effectiveResourceName = resourceName ?? sourceName, effectiveResource = options.resource ?? source;
  if (!effectiveResourceName && !effectiveResource)
    return rest;
  const hasNameKey = Object.hasOwn(options, "resourceName") || Object.hasOwn(options, "sourceName"), hasResourceKey = Object.hasOwn(options, "resource") || Object.hasOwn(options, "source");
  if (hasNameKey && hasResourceKey)
    throw new Error(`Resource name ${JSON.stringify(effectiveResourceName)} and resource ${JSON.stringify(effectiveResource)} cannot be used together.`);
  let resolvedResource;
  if (effectiveResource && (resolvedResource = effectiveResource), effectiveResourceName && !Object.hasOwn(resources, effectiveResourceName))
    throw new Error(`There's no resource named ${JSON.stringify(effectiveResourceName)} in context. Please use <ResourceProvider>.`);
  return effectiveResourceName && resources[effectiveResourceName] && (resolvedResource = resources[effectiveResourceName]), {
    ...rest,
    resource: resolvedResource
  };
}
function useNormalizedResourceOptions(options) {
  const $ = c(3), resources = useContext(ResourcesContext);
  let t0;
  return $[0] !== options || $[1] !== resources ? (t0 = normalizeResourceOptions(options, resources), $[0] = options, $[1] = resources, $[2] = t0) : t0 = $[2], t0;
}
function useResourceIdFromDocumentHandle(documentHandle) {
  const $ = c(3), options = useNormalizedResourceOptions(documentHandle), {
    projectId,
    dataset,
    resource
  } = options;
  let resourceId = "", resourceType;
  projectId && dataset && (resourceId = `${projectId}.${dataset}`), resource && (isDatasetResource(resource) ? (resourceId = `${resource.projectId}.${resource.dataset}`, resourceType = void 0) : isMediaLibraryResource(resource) ? (resourceId = resource.mediaLibraryId, resourceType = "media-library") : isCanvasResource(resource) && (resourceId = resource.canvasId, resourceType = "canvas"));
  let t0;
  return $[0] !== resourceId || $[1] !== resourceType ? (t0 = {
    id: resourceId,
    type: resourceType
  }, $[0] = resourceId, $[1] = resourceType, $[2] = t0) : t0 = $[2], t0;
}
function useDispatchIntent(params) {
  const {
    action,
    intentId,
    documentHandle,
    parameters
  } = params, {
    sendMessage
  } = useWindowConnection({
    name: SDK_NODE_NAME,
    connectTo: SDK_CHANNEL_NAME
  }), resource = useResourceIdFromDocumentHandle(documentHandle);
  return {
    dispatchIntent: useCallback(() => {
      try {
        if (!action && !intentId)
          throw new Error("useDispatchIntent: Either `action` or `intentId` must be provided.");
        if (action && intentId && console.warn("useDispatchIntent: Both `action` and `intentId` were provided. Using `intentId` and ignoring `action`."), !resource.id)
          throw new Error("useDispatchIntent: Unable to determine resource. Either `resource`, `resourceName`, or both `projectId` and `dataset` must be provided in documentHandle.");
        const message = {
          type: "dashboard/v1/events/intents/dispatch-intent",
          data: {
            ...action && !intentId ? {
              action
            } : {},
            ...intentId ? {
              intentId
            } : {},
            document: {
              id: documentHandle.documentId,
              type: documentHandle.documentType
            },
            resource: {
              id: resource.id,
              ...resource.type ? {
                type: resource.type
              } : {}
            },
            ...parameters && Object.keys(parameters).length > 0 ? {
              parameters
            } : {}
          }
        };
        sendMessage(message.type, message.data);
      } catch (error) {
        throw console.error("Failed to dispatch intent:", error), error;
      }
    }, [action, intentId, documentHandle, parameters, sendMessage, resource.id, resource.type])
  };
}
function useManageFavorite({
  documentId,
  documentType,
  projectId: paramProjectId,
  dataset: paramDataset,
  resourceId: paramResourceId,
  resourceType,
  schemaName
}) {
  const {
    fetch
  } = useWindowConnection({
    name: SDK_NODE_NAME,
    connectTo: SDK_CHANNEL_NAME
  }), instance = useSanityInstance(), {
    config
  } = instance, instanceProjectId = config?.projectId, instanceDataset = config?.dataset, projectId = paramProjectId ?? instanceProjectId, dataset = paramDataset ?? instanceDataset;
  if (resourceType === "studio" && (!projectId || !dataset))
    throw new Error("projectId and dataset are required for studio resources");
  const resourceId = resourceType === "studio" && !paramResourceId ? `${projectId}.${dataset}` : paramResourceId;
  if (!resourceId)
    throw new Error("resourceId is required for media-library and canvas resources");
  const context = useMemo(() => ({
    documentId,
    documentType,
    resourceId,
    resourceType,
    schemaName
  }), [documentId, documentType, resourceId, resourceType, schemaName]), favoriteState = getFavoritesState(instance, context), isFavorited = useSyncExternalStore(favoriteState.subscribe, favoriteState.getCurrent)?.isFavorited ?? !1, handleFavoriteAction = useCallback(async (action) => {
    if (!(!fetch || !documentId || !documentType || !resourceType))
      try {
        const payload = {
          eventType: action,
          document: {
            id: documentId,
            type: documentType,
            resource: {
              id: resourceId,
              type: resourceType,
              ...schemaName ? {
                schemaName
              } : {}
            }
          }
        };
        (await fetch("dashboard/v1/events/favorite/mutate", payload)).success && await resolveFavoritesState(instance, context);
      } catch (err) {
        throw console.error(`Failed to ${action === "added" ? "favorite" : "unfavorite"} document:`, err), err;
      }
  }, [fetch, documentId, documentType, resourceId, resourceType, schemaName, instance, context]), favorite = useCallback(() => handleFavoriteAction("added"), [handleFavoriteAction]), unfavorite = useCallback(() => handleFavoriteAction("removed"), [handleFavoriteAction]);
  return {
    favorite,
    unfavorite,
    isFavorited
  };
}
function useStudioWorkspacesByProjectIdDataset() {
  const $ = c(8);
  let t0;
  $[0] === Symbol.for("react.memo_cache_sentinel") ? (t0 = {}, $[0] = t0) : t0 = $[0];
  const [workspacesByProjectIdAndDataset, setWorkspacesByProjectIdAndDataset] = useState(t0), [error, setError] = useState(null);
  let t1;
  $[1] === Symbol.for("react.memo_cache_sentinel") ? (t1 = {
    name: SDK_NODE_NAME,
    connectTo: SDK_CHANNEL_NAME
  }, $[1] = t1) : t1 = $[1];
  const {
    fetch
  } = useWindowConnection(t1);
  let t2, t3;
  $[2] !== fetch ? (t2 = () => {
    if (!fetch)
      return;
    const fetchWorkspaces = async function(signal) {
      try {
        const data = await fetch("dashboard/v1/context", void 0, {
          signal
        }), workspaceMap = {}, noProjectIdAndDataset = [];
        data.context.availableResources.forEach((resource) => {
          if (resource.type !== "studio")
            return;
          if (!resource.projectId || !resource.dataset) {
            noProjectIdAndDataset.push(resource);
            return;
          }
          const key = `${resource.projectId}:${resource.dataset}`;
          workspaceMap[key] || (workspaceMap[key] = []), workspaceMap[key].push(resource);
        }), noProjectIdAndDataset.length > 0 && (workspaceMap["NO_PROJECT_ID:NO_DATASET"] = noProjectIdAndDataset), setWorkspacesByProjectIdAndDataset(workspaceMap), setError(null);
      } catch (t42) {
        const err = t42;
        if (err instanceof Error) {
          if (err.name === "AbortError")
            return;
          setError("Failed to fetch workspaces");
        }
      }
    }, controller = new AbortController();
    return fetchWorkspaces(controller.signal), () => {
      controller.abort();
    };
  }, t3 = [fetch], $[2] = fetch, $[3] = t2, $[4] = t3) : (t2 = $[3], t3 = $[4]), useEffect(t2, t3);
  let t4;
  return $[5] !== error || $[6] !== workspacesByProjectIdAndDataset ? (t4 = {
    workspacesByProjectIdAndDataset,
    error
  }, $[5] = error, $[6] = workspacesByProjectIdAndDataset, $[7] = t4) : t4 = $[7], t4;
}
function useNavigateToStudioDocument(documentHandle, preferredStudioUrl) {
  const $ = c(8), {
    workspacesByProjectIdAndDataset
  } = useStudioWorkspacesByProjectIdDataset();
  let t0;
  $[0] === Symbol.for("react.memo_cache_sentinel") ? (t0 = {
    name: SDK_NODE_NAME,
    connectTo: SDK_CHANNEL_NAME
  }, $[0] = t0) : t0 = $[0];
  const {
    sendMessage
  } = useWindowConnection(t0);
  let t1;
  $[1] !== documentHandle || $[2] !== preferredStudioUrl || $[3] !== sendMessage || $[4] !== workspacesByProjectIdAndDataset ? (t1 = () => {
    const {
      projectId,
      dataset
    } = documentHandle;
    if (!projectId || !dataset) {
      console.warn("Project ID and dataset are required to navigate to a studio document");
      return;
    }
    let workspace;
    if (preferredStudioUrl)
      workspace = [...workspacesByProjectIdAndDataset[`${projectId}:${dataset}`] || [], ...workspacesByProjectIdAndDataset["NO_PROJECT_ID:NO_DATASET"] || []].find((w) => w.url === preferredStudioUrl);
    else {
      const workspaces = workspacesByProjectIdAndDataset[`${projectId}:${dataset}`];
      workspaces?.length > 1 && (console.warn("Multiple workspaces found for document and no preferred studio url", documentHandle), console.warn("Using the first one", workspaces[0])), workspace = workspaces?.[0];
    }
    if (!workspace) {
      console.warn(`No workspace found for document with projectId: ${projectId} and dataset: ${dataset}${preferredStudioUrl ? ` or with preferred studio url: ${preferredStudioUrl}` : ""}`);
      return;
    }
    const message = {
      type: "dashboard/v1/bridge/navigate-to-resource",
      data: {
        resourceId: workspace.id,
        resourceType: "studio",
        path: `/intent/edit/id=${documentHandle.documentId};type=${documentHandle.documentType}`
      }
    };
    sendMessage(message.type, message.data);
  }, $[1] = documentHandle, $[2] = preferredStudioUrl, $[3] = sendMessage, $[4] = workspacesByProjectIdAndDataset, $[5] = t1) : t1 = $[5];
  const navigateToStudioDocument = t1;
  let t2;
  return $[6] !== navigateToStudioDocument ? (t2 = {
    navigateToStudioDocument
  }, $[6] = navigateToStudioDocument, $[7] = t2) : t2 = $[7], t2;
}
function useRecordDocumentHistoryEvent(t0) {
  const $ = c(10), {
    documentId,
    documentType,
    resourceType,
    resourceId,
    schemaName
  } = t0;
  let t1;
  $[0] === Symbol.for("react.memo_cache_sentinel") ? (t1 = {
    name: SDK_NODE_NAME,
    connectTo: SDK_CHANNEL_NAME
  }, $[0] = t1) : t1 = $[0];
  const {
    sendMessage
  } = useWindowConnection(t1);
  if (resourceType !== "studio" && !resourceId)
    throw new Error("resourceId is required for media-library and canvas resources");
  let t2;
  $[1] !== documentId || $[2] !== documentType || $[3] !== resourceId || $[4] !== resourceType || $[5] !== schemaName || $[6] !== sendMessage ? (t2 = (eventType) => {
    try {
      const message = {
        type: "dashboard/v1/events/history",
        data: {
          eventType,
          document: {
            id: documentId,
            type: documentType,
            resource: {
              id: resourceId,
              type: resourceType,
              schemaName
            }
          }
        }
      };
      sendMessage(message.type, message.data);
    } catch (t32) {
      const error = t32;
      throw console.error("Failed to record history event:", error), error;
    }
  }, $[1] = documentId, $[2] = documentType, $[3] = resourceId, $[4] = resourceType, $[5] = schemaName, $[6] = sendMessage, $[7] = t2) : t2 = $[7];
  const recordEvent = t2;
  let t3;
  return $[8] !== recordEvent ? (t3 = {
    recordEvent
  }, $[8] = recordEvent, $[9] = t3) : t3 = $[9], t3;
}
function resolveAppTitle(resource) {
  return resource.manifest?.title ?? resource.activeDeployment?.manifest?.title ?? resource.title;
}
function useWindowTitle(viewTitle) {
  const $ = c(8), [appTitle, setAppTitle] = useState(null);
  let t0;
  $[0] === Symbol.for("react.memo_cache_sentinel") ? (t0 = {
    name: SDK_NODE_NAME,
    connectTo: SDK_CHANNEL_NAME
  }, $[0] = t0) : t0 = $[0];
  const {
    fetch
  } = useWindowConnection(t0);
  let t1, t2;
  $[1] !== fetch ? (t1 = () => {
    if (!fetch)
      return;
    const controller = new AbortController();
    return (async function(signal) {
      try {
        const data = await fetch("dashboard/v1/context", void 0, {
          signal
        }), title = resolveAppTitle(data.context.resource);
        title && setAppTitle(title);
      } catch (t32) {
        const err = t32;
        if (err instanceof Error && err.name === "AbortError")
          return;
        console.error("Failed to fetch app title from dashboard context:", err);
      }
    })(controller.signal), () => {
      controller.abort();
    };
  }, t2 = [fetch], $[1] = fetch, $[2] = t1, $[3] = t2) : (t1 = $[2], t2 = $[3]), useEffect(t1, t2);
  let t3, t4;
  $[4] !== appTitle || $[5] !== viewTitle ? (t3 = () => {
    if (!appTitle)
      return;
    const previous = document.title;
    return document.title = viewTitle ? `${viewTitle} | ${appTitle}` : appTitle, () => {
      document.title = previous;
    };
  }, t4 = [viewTitle, appTitle], $[4] = appTitle, $[5] = viewTitle, $[6] = t3, $[7] = t4) : (t3 = $[6], t4 = $[7]), useEffect(t3, t4);
}
const useDatasets = createStateSourceHook({
  getState: getDatasetsState,
  shouldSuspend: (instance, projectHandle) => (
    // remove `undefined` since we're suspending when that is the case
    getDatasetsState(instance, projectHandle).getCurrent() === void 0
  ),
  suspender: resolveDatasets,
  getConfig: identity
}), useApplyDocumentActions = () => {
  const $ = c(3), instance = useSanityInstance(), resources = useContext(ResourcesContext);
  let t0;
  return $[0] !== instance || $[1] !== resources ? (t0 = (actionOrActions, options) => {
    const actions = Array.isArray(actionOrActions) ? actionOrActions : [actionOrActions], normalizedOptions = options ? normalizeResourceOptions(options, resources) : void 0;
    let projectId, dataset, resource;
    for (const action of actions) {
      if (action.projectId) {
        if (resource)
          throw new Error(`Mismatches between projectId/dataset options and resource in actions. Found projectId "${action.projectId}" and dataset "${action.dataset}" but expected resource "${resource}".`);
        if (projectId || (projectId = action.projectId), action.projectId !== projectId)
          throw new Error(`Mismatched project IDs found in actions. All actions must belong to the same project. Found "${action.projectId}" but expected "${projectId}".`);
        if (action.dataset && (dataset || (dataset = action.dataset), action.dataset !== dataset))
          throw new Error(`Mismatched datasets found in actions. All actions must belong to the same dataset. Found "${action.dataset}" but expected "${dataset}".`);
      }
      if (action.resource) {
        if (resource || (resource = action.resource), action.resource !== resource)
          throw new Error(`Mismatched resources found in actions. All actions must belong to the same resource. Found "${action.resource}" but expected "${resource}".`);
        if (projectId || dataset)
          throw new Error(`Mismatches between projectId/dataset options and resource in actions. Found "${action.resource}" but expected project "${projectId}" and dataset "${dataset}".`);
      }
    }
    if (projectId || dataset) {
      const actualInstance = instance.match({
        projectId,
        dataset
      });
      if (!actualInstance)
        throw new Error(`Could not find a matching Sanity instance for the requested action: ${JSON.stringify({
          projectId,
          dataset
        }, null, 2)}.
  Please ensure there is a ResourceProvider component with a matching configuration in the component hierarchy.`);
      return applyDocumentActions(actualInstance, {
        actions,
        resource,
        ...normalizedOptions
      });
    }
    return applyDocumentActions(instance, {
      actions,
      resource,
      ...normalizedOptions
    });
  }, $[0] = instance, $[1] = resources, $[2] = t0) : t0 = $[2], t0;
};
function useTrackHookUsage(hookName) {
  const instance = useSanityInstance(), tracked = useRef(null);
  tracked.current === null && (tracked.current = !0, trackHookMounted(instance, hookName));
}
function trackHookUsage(instance, hookName) {
  trackHookMounted(instance, hookName);
}
const useDocumentValue = createStateSourceHook({
  // Pass options directly to getDocumentState
  getState: (instance, options) => getDocumentState(instance, options),
  // Pass options directly to getDocumentState for checking current value
  shouldSuspend: (instance, {
    path: _path,
    ...options
  }) => getDocumentState(instance, options).getCurrent() === void 0,
  // Extract handle part for resolveDocument
  suspender: (instance, options) => resolveDocument(instance, options),
  getConfig: identity
}), wrapHookWithData = (useValue) => {
  function useHook(...params) {
    return {
      data: useValue(...params)
    };
  }
  return useHook;
}, useDocument = wrapHookWithData((options) => {
  useTrackHookUsage("useDocument");
  const normalizedOptions = useNormalizedResourceOptions(options);
  return useDocumentValue(normalizedOptions);
});
function useDocumentEvent(options) {
  const $ = c(10);
  useTrackHookUsage("useDocumentEvent");
  const normalizedOptions = useNormalizedResourceOptions(options);
  let datasetHandle, onEvent;
  $[0] !== normalizedOptions ? ({
    onEvent,
    ...datasetHandle
  } = normalizedOptions, $[0] = normalizedOptions, $[1] = datasetHandle, $[2] = onEvent) : (datasetHandle = $[1], onEvent = $[2]);
  const ref = useRef(onEvent);
  let t0;
  $[3] !== onEvent ? (t0 = () => {
    ref.current = onEvent;
  }, $[3] = onEvent, $[4] = t0) : t0 = $[4], useInsertionEffect(t0);
  let t1;
  $[5] === Symbol.for("react.memo_cache_sentinel") ? (t1 = (documentEvent) => ref.current(documentEvent), $[5] = t1) : t1 = $[5];
  const stableHandler = t1, instance = useSanityInstance(datasetHandle);
  let t2, t3;
  $[6] !== datasetHandle.resource || $[7] !== instance ? (t2 = () => subscribeDocumentEvents(instance, {
    eventHandler: stableHandler,
    resource: datasetHandle.resource
  }), t3 = [instance, datasetHandle.resource, stableHandler], $[6] = datasetHandle.resource, $[7] = instance, $[8] = t2, $[9] = t3) : (t2 = $[8], t3 = $[9]), useEffect(t2, t3);
}
function useDocumentPermissions(actionOrActions) {
  const $ = c(15);
  let t0, t1;
  $[0] !== actionOrActions ? (t1 = Array.isArray(actionOrActions) ? actionOrActions : [actionOrActions], $[0] = actionOrActions, $[1] = t1) : t1 = $[1], t0 = t1;
  const actions = t0;
  let projectId, dataset, resource;
  if ($[2] !== actions || $[3] !== dataset || $[4] !== projectId || $[5] !== resource) {
    for (const action of actions) {
      if (action.projectId) {
        if (resource)
          throw new Error(`Mismatches between projectId/dataset options and resource in actions. Found projectId "${action.projectId}" and dataset "${action.dataset}" but expected resource "${resource}".`);
        if (projectId || (projectId = action.projectId), action.projectId !== projectId)
          throw new Error(`Mismatched project IDs found in actions. All actions must belong to the same project. Found "${action.projectId}" but expected "${projectId}".`);
        if (action.dataset && (dataset || (dataset = action.dataset), action.dataset !== dataset))
          throw new Error(`Mismatched datasets found in actions. All actions must belong to the same dataset. Found "${action.dataset}" but expected "${dataset}".`);
      }
      if (action.resource) {
        if (resource || (resource = action.resource), action.resource !== resource)
          throw new Error(`Mismatched resources found in actions. All actions must belong to the same resource. Found "${action.resource}" but expected "${resource}".`);
        if (projectId || dataset)
          throw new Error(`Mismatches between projectId/dataset options and resource in actions. Found "${action.resource}" but expected project "${projectId}" and dataset "${dataset}".`);
      }
    }
    $[2] = actions, $[3] = dataset, $[4] = projectId, $[5] = resource, $[6] = projectId, $[7] = dataset, $[8] = resource;
  } else
    projectId = $[6], dataset = $[7], resource = $[8];
  let t2;
  $[9] !== dataset || $[10] !== projectId ? (t2 = {
    projectId,
    dataset
  }, $[9] = dataset, $[10] = projectId, $[11] = t2) : t2 = $[11];
  const instance = useSanityInstance(t2);
  if (trackHookUsage(instance, "useDocumentPermissions"), getPermissionsState(instance, {
    actions
  }).getCurrent() === void 0)
    throw firstValueFrom(getPermissionsState(instance, {
      actions
    }).observable.pipe(filter(_temp$2)));
  let t3, t4;
  $[12] !== actions || $[13] !== instance ? (t4 = getPermissionsState(instance, {
    actions
  }), $[12] = actions, $[13] = instance, $[14] = t4) : t4 = $[14], t3 = t4;
  const {
    subscribe,
    getCurrent
  } = t3;
  return useSyncExternalStore(subscribe, getCurrent);
}
function _temp$2(result) {
  return result !== void 0;
}
const useDocumentSyncStatusValue = createStateSourceHook({
  getState: getDocumentSyncStatus,
  shouldSuspend: (instance, doc) => getDocumentSyncStatus(instance, doc).getCurrent() === void 0,
  suspender: (instance, doc) => resolveDocument(instance, doc),
  getConfig: identity
}), useDocumentSyncStatus = (options) => {
  const normalizedOptions = useNormalizedResourceOptions(options);
  return useDocumentSyncStatusValue(normalizedOptions);
}, ignoredKeys = ["_id", "_type", "_createdAt", "_updatedAt", "_rev"];
function useEditDocument(t0) {
  const $ = c(8);
  let doc, path;
  $[0] !== t0 ? ({
    path,
    ...doc
  } = t0, $[0] = t0, $[1] = doc, $[2] = path) : (doc = $[1], path = $[2]);
  const instance = useSanityInstance(doc);
  trackHookUsage(instance, "useEditDocument");
  const normalizedDoc = useNormalizedResourceOptions(doc), apply = useApplyDocumentActions();
  if (getDocumentState(instance, normalizedDoc).getCurrent() === void 0)
    throw resolveDocument(instance, normalizedDoc);
  let t1;
  return $[3] !== apply || $[4] !== instance || $[5] !== normalizedDoc || $[6] !== path ? (t1 = (updater) => {
    const currentPath = path;
    if (currentPath) {
      const currentValue = getDocumentState(instance, {
        ...normalizedDoc,
        path
      }).getCurrent(), nextValue = typeof updater == "function" ? updater(currentValue) : updater;
      return apply(editDocument(normalizedDoc, {
        set: {
          [currentPath]: nextValue
        }
      }));
    }
    const current = getDocumentState(instance, {
      ...normalizedDoc,
      path
    }).getCurrent(), nextValue_0 = typeof updater == "function" ? updater(current) : updater;
    if (typeof nextValue_0 != "object" || !nextValue_0)
      throw new Error("No path was provided to `useEditDocument` and the value provided was not a document object.");
    const editActions = Object.keys({
      ...current,
      ...nextValue_0
    }).filter(_temp$1).filter((key_0) => current?.[key_0] !== nextValue_0[key_0]).map((key_1) => key_1 in nextValue_0 ? editDocument(normalizedDoc, {
      set: {
        [key_1]: nextValue_0[key_1]
      }
    }) : editDocument(normalizedDoc, {
      unset: [key_1]
    }));
    return apply(editActions);
  }, $[3] = apply, $[4] = instance, $[5] = normalizedDoc, $[6] = path, $[7] = t1) : t1 = $[7], t1;
}
function _temp$1(key) {
  return !ignoredKeys.includes(key);
}
function useQuery(options) {
  const instance = useSanityInstance(options);
  trackHookUsage(instance, "useQuery");
  const normalized = useNormalizedResourceOptions(options), [isPending, startTransition] = useTransition(), queryKey = getQueryKey(normalized), [deferredQueryKey, setDeferredQueryKey] = useState(queryKey), ref = useRef(new AbortController());
  useEffect(() => {
    queryKey !== deferredQueryKey && startTransition(() => {
      ref && !ref.current.signal.aborted && (ref.current.abort(), ref.current = new AbortController()), setDeferredQueryKey(queryKey);
    });
  }, [deferredQueryKey, queryKey]);
  const {
    getCurrent,
    subscribe
  } = useMemo(() => {
    const deferred = parseQueryKey(deferredQueryKey);
    return getQueryState(instance, deferred);
  }, [instance, deferredQueryKey]);
  if (getCurrent() === void 0) {
    const currentSignal = ref.current.signal, deferred_0 = parseQueryKey(deferredQueryKey);
    throw resolveQuery(instance, {
      ...deferred_0,
      signal: currentSignal
    });
  }
  const data = useSyncExternalStore(subscribe, getCurrent);
  return useMemo(() => ({
    data,
    isPending
  }), [data, isPending]);
}
const DEFAULT_BATCH_SIZE = 25;
function useDocuments({
  batchSize = DEFAULT_BATCH_SIZE,
  params,
  search,
  filter: filter2,
  orderings,
  documentType,
  ...options
}) {
  useTrackHookUsage("useDocuments");
  const instance = useSanityInstance(options), [limit, setLimit] = useState(batchSize), documentTypes = useMemo(() => (Array.isArray(documentType) ? documentType : [documentType]).filter((i) => typeof i == "string"), [documentType]), key = JSON.stringify({
    filter: filter2,
    search,
    params,
    orderings,
    batchSize,
    types: documentTypes,
    ...options
  }), [prevKey, setPrevKey] = useState(key);
  prevKey !== key && (setPrevKey(key), setLimit(batchSize));
  const filterClause = useMemo(() => {
    const conditions = [], trimmedSearch = search?.trim();
    if (trimmedSearch) {
      const searchFilter = createGroqSearchFilter(trimmedSearch);
      searchFilter && conditions.push(searchFilter);
    }
    return documentTypes?.length && conditions.push("(_type in $__types)"), filter2 && conditions.push(`(${filter2})`), conditions.length ? `[${conditions.join(" && ")}]` : "";
  }, [filter2, search, documentTypes]), orderClause = orderings ? `| order(${orderings.map((ordering) => [ordering.field, ordering.direction.toLowerCase()].map((str) => str.trim()).filter(Boolean).join(" ")).join(",")})` : "", dataQuery = `*${filterClause}${orderClause}[0...${limit}]{"documentId":_id,"documentType":_type,...$__handle}`, countQuery = `count(*${filterClause})`, {
    data: {
      count,
      data
    },
    isPending
  } = useQuery({
    ...options,
    query: `{"count":${countQuery},"data":${dataQuery}}`,
    params: {
      ...params,
      // these are passed back to the user as part of each document handle
      __handle: {
        projectId: options.projectId ?? instance.config.projectId,
        dataset: options.dataset ?? instance.config.dataset,
        perspective: options.perspective ?? instance.config.perspective
      },
      __types: documentTypes
    }
  }), hasMore = data.length < count, loadMore = useCallback(() => {
    setLimit((prev) => Math.min(prev + batchSize, count));
  }, [count, batchSize]);
  return useMemo(() => ({
    data,
    hasMore,
    count,
    isPending,
    loadMore
  }), [count, data, hasMore, isPending, loadMore]);
}
function usePaginatedDocuments(t0) {
  const $ = c(56);
  let documentType, options, orderings, search, t1, t2, t3;
  $[0] !== t0 ? ({
    documentType,
    filter: t1,
    pageSize: t2,
    params: t3,
    orderings,
    search,
    ...options
  } = t0, $[0] = t0, $[1] = documentType, $[2] = options, $[3] = orderings, $[4] = search, $[5] = t1, $[6] = t2, $[7] = t3) : (documentType = $[1], options = $[2], orderings = $[3], search = $[4], t1 = $[5], t2 = $[6], t3 = $[7]);
  const filter2 = t1 === void 0 ? "" : t1, pageSize = t2 === void 0 ? 25 : t2;
  let t4;
  $[8] !== t3 ? (t4 = t3 === void 0 ? {} : t3, $[8] = t3, $[9] = t4) : t4 = $[9];
  const params = t4;
  useTrackHookUsage("usePaginatedDocuments");
  const instance = useSanityInstance(options), [pageIndex, setPageIndex] = useState(0);
  let t5;
  $[10] !== filter2 || $[11] !== orderings || $[12] !== pageSize || $[13] !== params || $[14] !== search ? (t5 = JSON.stringify({
    filter: filter2,
    search,
    params,
    orderings,
    pageSize
  }), $[10] = filter2, $[11] = orderings, $[12] = pageSize, $[13] = params, $[14] = search, $[15] = t5) : t5 = $[15];
  const key = t5, [prevKey, setPrevKey] = useState(key);
  prevKey !== key && (setPrevKey(key), setPageIndex(0));
  const startIndex = pageIndex * pageSize, endIndex = (pageIndex + 1) * pageSize;
  let t6;
  $[16] !== documentType ? (t6 = Array.isArray(documentType) ? documentType : [documentType], $[16] = documentType, $[17] = t6) : t6 = $[17];
  let t7;
  $[18] !== t6 ? (t7 = t6.filter(_temp), $[18] = t6, $[19] = t7) : t7 = $[19];
  const documentTypes = t7;
  let t8;
  const conditions = [], trimmedSearch = search?.trim();
  if (trimmedSearch) {
    const searchFilter = createGroqSearchFilter(trimmedSearch);
    searchFilter && conditions.push(searchFilter);
  }
  documentTypes?.length && conditions.push("(_type in $__types)"), filter2 && conditions.push(`(${filter2})`), t8 = conditions.length ? `[${conditions.join(" && ")}]` : "";
  const filterClause = t8, orderClause = orderings ? `| order(${orderings.map(_temp3).join(",")})` : "", dataQuery = `*${filterClause}${orderClause}[${startIndex}...${endIndex}]{"documentId":_id,"documentType":_type,...$__handle}`, countQuery = `count(*${filterClause})`, t9 = `{"data":${dataQuery},"count":${countQuery}}`, t10 = options.projectId ?? instance.config.projectId, t11 = options.dataset ?? instance.config.dataset, t12 = options.perspective ?? instance.config.perspective;
  let t13;
  $[20] !== t10 || $[21] !== t11 || $[22] !== t12 ? (t13 = {
    projectId: t10,
    dataset: t11,
    perspective: t12
  }, $[20] = t10, $[21] = t11, $[22] = t12, $[23] = t13) : t13 = $[23];
  let t14;
  $[24] !== documentTypes || $[25] !== params || $[26] !== t13 ? (t14 = {
    ...params,
    __types: documentTypes,
    __handle: t13
  }, $[24] = documentTypes, $[25] = params, $[26] = t13, $[27] = t14) : t14 = $[27];
  let t15;
  $[28] !== options || $[29] !== t14 || $[30] !== t9 ? (t15 = {
    ...options,
    query: t9,
    params: t14
  }, $[28] = options, $[29] = t14, $[30] = t9, $[31] = t15) : t15 = $[31];
  const {
    data: t16,
    isPending
  } = useQuery(t15), {
    data,
    count
  } = t16, totalPages = Math.ceil(count / pageSize), currentPage = pageIndex + 1;
  let t17;
  $[32] === Symbol.for("react.memo_cache_sentinel") ? (t17 = () => setPageIndex(0), $[32] = t17) : t17 = $[32];
  const firstPage = t17;
  let t18;
  $[33] === Symbol.for("react.memo_cache_sentinel") ? (t18 = () => setPageIndex(_temp4), $[33] = t18) : t18 = $[33];
  const previousPage = t18;
  let t19;
  $[34] !== totalPages ? (t19 = () => setPageIndex((prev_0) => Math.min(prev_0 + 1, totalPages - 1)), $[34] = totalPages, $[35] = t19) : t19 = $[35];
  const nextPage = t19;
  let t20;
  $[36] !== totalPages ? (t20 = () => setPageIndex(totalPages - 1), $[36] = totalPages, $[37] = t20) : t20 = $[37];
  const lastPage = t20;
  let t21;
  $[38] !== totalPages ? (t21 = (pageNumber) => {
    pageNumber < 1 || pageNumber > totalPages || setPageIndex(pageNumber - 1);
  }, $[38] = totalPages, $[39] = t21) : t21 = $[39];
  const goToPage = t21, hasFirstPage = pageIndex > 0, hasPreviousPage = pageIndex > 0, hasNextPage = pageIndex < totalPages - 1, hasLastPage = pageIndex < totalPages - 1;
  let t22;
  return $[40] !== count || $[41] !== currentPage || $[42] !== data || $[43] !== endIndex || $[44] !== goToPage || $[45] !== hasFirstPage || $[46] !== hasLastPage || $[47] !== hasNextPage || $[48] !== hasPreviousPage || $[49] !== isPending || $[50] !== lastPage || $[51] !== nextPage || $[52] !== pageSize || $[53] !== startIndex || $[54] !== totalPages ? (t22 = {
    data,
    isPending,
    pageSize,
    currentPage,
    totalPages,
    startIndex,
    endIndex,
    count,
    firstPage,
    hasFirstPage,
    previousPage,
    hasPreviousPage,
    nextPage,
    hasNextPage,
    lastPage,
    hasLastPage,
    goToPage
  }, $[40] = count, $[41] = currentPage, $[42] = data, $[43] = endIndex, $[44] = goToPage, $[45] = hasFirstPage, $[46] = hasLastPage, $[47] = hasNextPage, $[48] = hasPreviousPage, $[49] = isPending, $[50] = lastPage, $[51] = nextPage, $[52] = pageSize, $[53] = startIndex, $[54] = totalPages, $[55] = t22) : t22 = $[55], t22;
}
function _temp4(prev) {
  return Math.max(prev - 1, 0);
}
function _temp3(ordering) {
  return [ordering.field, ordering.direction.toLowerCase()].map(_temp2).filter(Boolean).join(" ");
}
function _temp2(str) {
  return str.trim();
}
function _temp(i) {
  return typeof i == "string";
}
function usePresence(t0) {
  const $ = c(14);
  let t1;
  $[0] !== t0 ? (t1 = t0 === void 0 ? {} : t0, $[0] = t0, $[1] = t1) : t1 = $[1];
  const normalizedOptions = useNormalizedResourceOptions(t1);
  if (normalizedOptions.resource && isMediaLibraryResource(normalizedOptions.resource))
    throw new Error("usePresence() does not support media library resources. Presence tracking requires a canvas or dataset resource.");
  const sanityInstance = useSanityInstance();
  trackHookUsage(sanityInstance, "usePresence");
  let t2, t3;
  $[2] !== normalizedOptions || $[3] !== sanityInstance ? (t3 = getPresence(sanityInstance, normalizedOptions), $[2] = normalizedOptions, $[3] = sanityInstance, $[4] = t3) : t3 = $[4], t2 = t3;
  const source = t2;
  let t4;
  $[5] !== source ? (t4 = (callback) => source.subscribe(callback), $[5] = source, $[6] = t4) : t4 = $[6];
  const subscribe = t4;
  let t5, t6;
  $[7] !== source ? (t5 = () => source.getCurrent(), t6 = () => source.getCurrent(), $[7] = source, $[8] = t5, $[9] = t6) : (t5 = $[8], t6 = $[9]);
  const locations = useSyncExternalStore(subscribe, t5, t6);
  let t7;
  $[10] !== locations ? (t7 = locations || [], $[10] = locations, $[11] = t7) : t7 = $[11];
  let t8;
  return $[12] !== t7 ? (t8 = {
    locations: t7
  }, $[12] = t7, $[13] = t8) : t8 = $[13], t8;
}
function useDocumentProjection({
  ref,
  projection,
  ...docHandle
}) {
  const instance = useSanityInstance(docHandle);
  trackHookUsage(instance, "useDocumentProjection");
  const normalizedProjection = useMemo(() => projection.trim(), [projection]), normalizedDocHandle = useNormalizedResourceOptions(docHandle), stateSource = useMemo(() => getProjectionState(instance, {
    ...normalizedDocHandle,
    projection: normalizedProjection
  }), [instance, normalizedDocHandle, normalizedProjection]);
  if (stateSource.getCurrent()?.data === null)
    throw resolveProjection(instance, {
      ...normalizedDocHandle,
      projection: normalizedProjection
    });
  const subscribe = useCallback((onStoreChanged) => {
    const subscription = new Observable((observer) => {
      if (typeof IntersectionObserver > "u" || typeof HTMLElement > "u") {
        observer.next(!0);
        return;
      }
      const intersectionObserver = new IntersectionObserver(([entry]) => observer.next(entry.isIntersecting), {
        rootMargin: "0px",
        threshold: 0
      });
      return ref?.current && ref.current instanceof HTMLElement ? intersectionObserver.observe(ref.current) : observer.next(!0), () => intersectionObserver.disconnect();
    }).pipe(startWith(!1), distinctUntilChanged(), switchMap((isVisible) => isVisible ? new Observable((obs) => stateSource.subscribe(() => obs.next())) : EMPTY)).subscribe({
      next: onStoreChanged
    });
    return () => subscription.unsubscribe();
  }, [stateSource, ref]);
  return useSyncExternalStore(subscribe, stateSource.getCurrent);
}
function useDocumentPreview(t0) {
  const $ = c(13);
  let docHandle, ref;
  $[0] !== t0 ? ({
    ref,
    ...docHandle
  } = t0, $[0] = t0, $[1] = docHandle, $[2] = ref) : (docHandle = $[1], ref = $[2]);
  const instance = useSanityInstance(docHandle);
  trackHookUsage(instance, "useDocumentPreview");
  const normalizedDocHandle = useNormalizedResourceOptions(docHandle);
  let t1;
  $[3] !== normalizedDocHandle || $[4] !== ref ? (t1 = {
    ...normalizedDocHandle,
    projection: PREVIEW_PROJECTION,
    ref
  }, $[3] = normalizedDocHandle, $[4] = ref, $[5] = t1) : t1 = $[5];
  const projectionResult = useDocumentProjection(t1);
  let t2, t3;
  $[6] !== instance || $[7] !== normalizedDocHandle.resource || $[8] !== projectionResult.data ? (t3 = transformProjectionToPreview(instance, projectionResult.data, normalizedDocHandle.resource), $[6] = instance, $[7] = normalizedDocHandle.resource, $[8] = projectionResult.data, $[9] = t3) : t3 = $[9], t2 = t3;
  const previewValue = t2;
  let t4;
  return $[10] !== previewValue || $[11] !== projectionResult.isPending ? (t4 = {
    data: previewValue,
    isPending: projectionResult.isPending
  }, $[10] = previewValue, $[11] = projectionResult.isPending, $[12] = t4) : t4 = $[12], t4;
}
const useProject = createStateSourceHook({
  // remove `undefined` since we're suspending when that is the case
  getState: getProjectState,
  shouldSuspend: (instance, projectHandle) => getProjectState(instance, projectHandle).getCurrent() === void 0,
  suspender: resolveProject,
  getConfig: identity
}), useProjects = createStateSourceHook({
  getState: getProjectsState,
  shouldSuspend: (instance, options) => getProjectsState(instance, options).getCurrent() === void 0,
  suspender: resolveProjects
}), useActiveReleasesValue = createStateSourceHook({
  getState: getActiveReleasesState,
  shouldSuspend: (instance, options) => getActiveReleasesState(instance, options ?? {}).getCurrent() === void 0,
  suspender: (instance, options) => firstValueFrom(getActiveReleasesState(instance, options ?? {}).observable.pipe(filter(Boolean)))
}), useActiveReleases = (options) => {
  const $ = c(2);
  let t0;
  $[0] !== options ? (t0 = options ?? {}, $[0] = options, $[1] = t0) : t0 = $[1];
  const normalizedOptions = useNormalizedResourceOptions(t0);
  return useActiveReleasesValue(normalizedOptions);
}, usePerspectiveValue = createStateSourceHook({
  getState: getPerspectiveState,
  shouldSuspend: (instance, options) => getPerspectiveState(instance, options).getCurrent() === void 0,
  suspender: (instance, _options) => firstValueFrom(getPerspectiveState(instance, _options ?? {}).observable.pipe(filter(Boolean)))
}), usePerspective = (options) => {
  const $ = c(2);
  let t0;
  $[0] !== options ? (t0 = options ?? {}, $[0] = options, $[1] = t0) : t0 = $[1];
  const normalizedOptions = useNormalizedResourceOptions(t0);
  return usePerspectiveValue(normalizedOptions);
};
function useUser(options) {
  const instance = useSanityInstance(options);
  trackHookUsage(instance, "useUser");
  const [isPending, startTransition] = useTransition(), key = getUsersKey(instance, options), [deferredKey, setDeferredKey] = useState(key), deferred = useMemo(() => parseUsersKey(deferredKey), [deferredKey]), [ref, setRef] = useState(new AbortController());
  useEffect(() => {
    key !== deferredKey && startTransition(() => {
      ref.signal.aborted || (ref.abort(), setRef(new AbortController())), setDeferredKey(key);
    });
  }, [deferredKey, key, ref]);
  const {
    getCurrent,
    subscribe
  } = useMemo(() => getUsersState(instance, deferred), [instance, deferred]);
  if (getCurrent() === void 0)
    throw resolveUsers(instance, {
      ...deferred,
      signal: ref.signal
    });
  return {
    data: useSyncExternalStore(subscribe, getCurrent)?.data[0],
    isPending
  };
}
function useUsers(options) {
  const instance = useSanityInstance(options);
  trackHookUsage(instance, "useUsers");
  const [isPending, startTransition] = useTransition(), key = getUsersKey(instance, options), [deferredKey, setDeferredKey] = useState(key), deferred = useMemo(() => parseUsersKey(deferredKey), [deferredKey]), [ref, setRef] = useState(new AbortController());
  useEffect(() => {
    key !== deferredKey && startTransition(() => {
      ref.signal.aborted || (ref.abort(), setRef(new AbortController())), setDeferredKey(key);
    });
  }, [deferredKey, key, ref]);
  const {
    getCurrent,
    subscribe
  } = useMemo(() => getUsersState(instance, deferred), [instance, deferred]);
  if (getCurrent() === void 0)
    throw resolveUsers(instance, {
      ...deferred,
      signal: ref.signal
    });
  const {
    data,
    hasMore
  } = useSyncExternalStore(subscribe, getCurrent), loadMore = useCallback(() => {
    loadMoreUsers(instance, options);
  }, [instance, options]);
  return {
    data,
    hasMore,
    isPending,
    loadMore
  };
}
var version = "2.10.0";
function getEnv(key) {
  if (typeof import.meta < "u" && import.meta.env)
    return import.meta.env[key];
  if (typeof process < "u" && process.env)
    return process.env[key];
  if (typeof window < "u" && window.ENV)
    return window.ENV?.[key];
}
const REACT_SDK_VERSION = getEnv("PKG_VERSION") || `${version}-development`;
export {
  AuthBoundary,
  ComlinkTokenRefreshProvider,
  REACT_SDK_VERSION,
  ResourceProvider,
  SDKProvider,
  SDKStudioContext,
  SanityApp,
  SanityInstanceProvider,
  renderSanityApp,
  useActiveReleases,
  useAgentGenerate,
  useAgentPatch,
  useAgentPrompt,
  useAgentResourceContext,
  useAgentTransform,
  useAgentTranslate,
  useApplyDocumentActions,
  useAuthState,
  useAuthToken,
  useClient,
  useCurrentUser,
  useDashboardNavigate,
  useDashboardOrganizationId,
  useDatasets,
  useDispatchIntent,
  useDocument,
  useDocumentEvent,
  useDocumentPermissions,
  useDocumentPreview,
  useDocumentProjection,
  useDocumentSyncStatus,
  useDocuments,
  useEditDocument,
  useFrameConnection,
  useHandleAuthCallback,
  useLogOut,
  useLoginUrl,
  useManageFavorite,
  useNavigateToStudioDocument,
  usePaginatedDocuments,
  usePerspective,
  usePresence,
  useProject,
  useProjects,
  useQuery,
  useRecordDocumentHistoryEvent,
  useSanityInstance,
  useStudioWorkspacesByProjectIdDataset,
  useUser,
  useUsers,
  useVerifyOrgProjects,
  useWindowConnection,
  useWindowTitle
};
