// @portabletext/plugin-sdk-value/dist/index.js
import { jsx } from "react/jsx-runtime";
import { c } from "react/compiler-runtime";
import { useEditor } from "@portabletext/editor";
import { diffValue } from "@sanity/diff-patch";
import { parsePath } from "@sanity/json-match";
import { useEditDocument, useSanityInstance, getDocumentState } from "@sanity/sdk-react";
import { useActorRef } from "@xstate/react";
import { fromCallback, setup } from "xstate";
const ARRAYIFY_ERROR_MESSAGE = "Unexpected path format from diffValue output. Please report this issue.";
function* getSegments(node) {
  node.base && (yield* getSegments(node.base)), node.segment.type !== "This" && (yield node.segment);
}
function isKeyPath(node) {
  return node.type !== "Path" || node.base || node.recursive || node.segment.type !== "Identifier" ? !1 : node.segment.name === "_key";
}
function arrayifyPath(pathExpr) {
  const node = parsePath(pathExpr);
  if (!node)
    return [];
  if (node.type !== "Path")
    throw new Error(ARRAYIFY_ERROR_MESSAGE);
  return Array.from(getSegments(node)).map((segment) => {
    if (segment.type === "Identifier")
      return segment.name;
    if (segment.type !== "Subscript")
      throw new Error(ARRAYIFY_ERROR_MESSAGE);
    if (segment.elements.length !== 1)
      throw new Error(ARRAYIFY_ERROR_MESSAGE);
    const [element] = segment.elements;
    if (element.type === "Number")
      return element.value;
    if (element.type !== "Comparison")
      throw new Error(ARRAYIFY_ERROR_MESSAGE);
    if (element.operator !== "==")
      throw new Error(ARRAYIFY_ERROR_MESSAGE);
    const keyPathNode = [element.left, element.right].find(isKeyPath);
    if (!keyPathNode)
      throw new Error(ARRAYIFY_ERROR_MESSAGE);
    const other = element.left === keyPathNode ? element.right : element.left;
    if (other.type !== "String")
      throw new Error(ARRAYIFY_ERROR_MESSAGE);
    return {
      _key: other.value
    };
  });
}
function convertPatches(patches) {
  return patches.flatMap((p) => Object.entries(p).flatMap(([type, values]) => {
    const origin = "remote";
    switch (type) {
      case "set":
      case "setIfMissing":
      case "diffMatchPatch":
      case "inc":
      case "dec":
        return Object.entries(values).map(([pathExpr, value]) => ({
          type,
          value,
          origin,
          path: arrayifyPath(pathExpr)
        }));
      case "unset":
        return Array.isArray(values) ? values.map(arrayifyPath).map((path) => ({
          type,
          origin,
          path
        })) : [];
      case "insert": {
        const {
          items,
          ...rest
        } = values, position = Object.keys(rest).at(0);
        if (!position)
          return [];
        const pathExpr = rest[position];
        return [{
          type,
          origin,
          position,
          path: arrayifyPath(pathExpr),
          items
        }];
      }
      default:
        return [];
    }
  }));
}
function applySync({
  editor,
  getRemoteValue
}) {
  const remoteValue = getRemoteValue();
  if (!remoteValue)
    return;
  const snapshot = editor.getSnapshot().context.value, patches = convertPatches(diffValue(snapshot, remoteValue));
  patches.length && editor.send({
    type: "patches",
    patches,
    snapshot
  });
}
const listenToEditor = fromCallback(({
  sendBack,
  input
}) => {
  const patchSubscription = input.editor.on("patch", () => {
    sendBack({
      type: "patch emitted"
    });
  }), mutationSubscription = input.editor.on("mutation", (event) => {
    sendBack({
      type: "mutation flushed",
      value: event.value
    });
  });
  return () => {
    patchSubscription.unsubscribe(), mutationSubscription.unsubscribe();
  };
}), listenToRemote = fromCallback(({
  sendBack,
  input
}) => input.onRemoteValueChange(() => {
  sendBack({
    type: "remote value changed"
  });
})), valueSyncMachine = setup({
  types: {
    context: {},
    input: {},
    events: {}
  },
  actions: {
    "send initial value": ({
      context
    }) => {
      context.editor.send({
        type: "update value",
        value: context.getRemoteValue() ?? []
      });
    },
    "push to remote": () => {
      throw new Error("push to remote must be provided via .provide()");
    },
    "apply sync": ({
      context
    }) => {
      applySync({
        editor: context.editor,
        getRemoteValue: context.getRemoteValue
      });
    },
    "defer then apply sync": ({
      context
    }) => {
      queueMicrotask(() => {
        applySync({
          editor: context.editor,
          getRemoteValue: context.getRemoteValue
        });
      });
    }
  },
  actors: {
    "listen to editor": listenToEditor,
    "listen to remote": listenToRemote
  }
}).createMachine({
  id: "value sync",
  context: ({
    input
  }) => ({
    editor: input.editor,
    getRemoteValue: input.getRemoteValue,
    onRemoteValueChange: input.onRemoteValueChange
  }),
  entry: ["send initial value"],
  invoke: [{
    src: "listen to editor",
    input: ({
      context
    }) => ({
      editor: context.editor
    })
  }, {
    src: "listen to remote",
    input: ({
      context
    }) => ({
      onRemoteValueChange: context.onRemoteValueChange
    })
  }],
  initial: "idle",
  states: {
    idle: {
      on: {
        "patch emitted": {
          target: "local write"
        },
        "remote value changed": {
          actions: ["apply sync"]
        }
      }
    },
    "local write": {
      on: {
        "patch emitted": {},
        "mutation flushed": {
          target: "pushing to remote",
          actions: ["push to remote"]
        },
        "remote value changed": {
          target: "pending sync"
        }
      }
    },
    "pushing to remote": {
      on: {
        "patch emitted": {
          target: "local write"
        },
        "remote value changed": {
          target: "idle"
        }
      }
    },
    "pending sync": {
      on: {
        "patch emitted": {},
        "mutation flushed": {
          target: "pushing to remote",
          actions: ["push to remote", "defer then apply sync"]
        },
        "remote value changed": {}
      }
    }
  }
});
function SDKValuePlugin(props) {
  const $ = c(9), {
    documentId,
    documentType,
    path
  } = props, setSdkValue = useEditDocument(props), instance = useSanityInstance(props);
  let t0;
  $[0] !== documentId || $[1] !== documentType || $[2] !== instance || $[3] !== path ? (t0 = getDocumentState(instance, {
    documentId,
    documentType,
    path
  }), $[0] = documentId, $[1] = documentType, $[2] = instance, $[3] = path, $[4] = t0) : t0 = $[4];
  const {
    getCurrent,
    subscribe
  } = t0;
  let t1;
  return $[5] !== getCurrent || $[6] !== setSdkValue || $[7] !== subscribe ? (t1 = /* @__PURE__ */ jsx(ValueSyncPlugin, { getRemoteValue: getCurrent, pushValue: setSdkValue, onRemoteValueChange: subscribe }), $[5] = getCurrent, $[6] = setSdkValue, $[7] = subscribe, $[8] = t1) : t1 = $[8], t1;
}
function ValueSyncPlugin(props) {
  const $ = c(6), {
    getRemoteValue,
    pushValue,
    onRemoteValueChange
  } = props, editor = useEditor();
  let t0;
  $[0] !== pushValue ? (t0 = valueSyncMachine.provide({
    actions: {
      "push to remote": (t12) => {
        const {
          context,
          event
        } = t12;
        event.type === "mutation flushed" && pushValue(event.value ?? context.editor.getSnapshot().context.value);
      }
    }
  }), $[0] = pushValue, $[1] = t0) : t0 = $[1];
  let t1;
  return $[2] !== editor || $[3] !== getRemoteValue || $[4] !== onRemoteValueChange ? (t1 = {
    input: {
      editor,
      getRemoteValue,
      onRemoteValueChange
    }
  }, $[2] = editor, $[3] = getRemoteValue, $[4] = onRemoteValueChange, $[5] = t1) : t1 = $[5], useActorRef(t0, t1), null;
}
export {
  SDKValuePlugin,
  ValueSyncPlugin
};
